Add integration tests for grafana deployment

ggriffiths opened the pull request:
**What this PR does / why we need it**:
Adds a few tests to the operator integration test suite for validating grafana deployments, svcs, configmaps, and airgapped support.

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

ggriffiths commented:
Ran the test suite a few times, works as expected:
```
--- PASS: TestGrafana (173.22s)
    --- PASS: TestGrafana/GrafanaInstall (173.15s)
    --- PASS: TestGrafana/GrafanaInstallAirgapped (0.07s)
PASS
==================================================
--- PASS:
    GrafanaInstall [XXX XXX]
    GrafanaInstallAirgapped [XXX XXX]
--- SKIP:
--- FAIL:
==================================================

DONE 3 tests in 173.491s
Tests passed
```

nikolaypopov commented on pkg/util/test/util.go:
```diff
@@ -2852,6 +2854,10 @@ func validateStorkSecurityEnvVar(cluster *corev1.StorageCluster, storkDeployment
 		return err
 	}
 
+	if err := ValidateGrafana(cluster); err != nil {
```
Why are we validating this as a part of `validateStorkSecurityEnvVar()` function?

nikolaypopov commented on pkg/util/test/util.go:
```diff
@@ -2928,6 +2934,171 @@ func ValidatePrometheus(pxImageList map[string]string, cluster *corev1.StorageCl
 	return nil
 }
 
+func ValidateGrafana(cluster *corev1.StorageCluster) error {
```
Can you add this as a part of `ValidateStorageCluster()` method

nikolaypopov commented on pkg/util/test/util.go:
```diff
@@ -2928,6 +2934,171 @@ func ValidatePrometheus(pxImageList map[string]string, cluster *corev1.StorageCl
+	return nil
+}
+
+func ValidateGrafanaImage(shouldBeInstalled bool) error {
```
This doesn't actually validate Grafana Image, we should do it though as it is part of release manifest versions URL

nikolaypopov commented on pkg/util/test/util.go:
```diff
@@ -2928,6 +2934,171 @@ func ValidatePrometheus(pxImageList map[string]string, cluster *corev1.StorageCl
+	return nil
+}
+
+func ValidateGrafanaDeploymentImage(image string) error {
```
Use this function as part of your `ValidateGrafanaDeployment()`, we need to validate images as a part of deployment validation, see example in the `ValidateStork()` or other components validation functions, also take a look at this function that does image validation inside containers https://github.com/libopenstorage/operator/blob/e4f90a46c0a2e28416d5acf44e953a5c42398010/pkg/util/test/util.go#L2577, this works with custom registries and trims image names accordingly using operator function `GetImageURN()`
NOTE: You need to validate images when they are part of the version image list https://install.portworx.com/3.0.0/version?kbver=1.24.0

nikolaypopov commented on pkg/util/test/util.go:
```diff
@@ -2928,6 +2934,171 @@ func ValidatePrometheus(pxImageList map[string]string, cluster *corev1.StorageCl
+	if err != nil {
+		return err
+	}
+	if opVersion.LessThanOrEqual(opVer23_8) {
```
This will skip validation even if your PX Operator version is 23.8, I thought you wanted to validate this starting with 23.8?

nikolaypopov reviewed (commented):
Some changes are needed here, let me know if you want to sync up

ggriffiths commented:
Tested again after addressing comments:
```
time="2023-07-25T19:53:35Z" level=info msg="Using version extracted from image name: 2.13.6"
time="2023-07-25T19:53:35Z" level=info msg="Update StorageCluster px-test-grafana in kube-system"
2023/07/25 19:53:35 grafana is not installed with image docker.io/grafana/grafana:8.5.27. number of pods: 0 Next retry in: 10s
time="2023-07-25T19:53:45Z" level=info msg="Grafana installed successfully with image docker.io/grafana/grafana:8.5.27"
--- PASS: TestGrafana (158.30s)
    --- PASS: TestGrafana/GrafanaInstall (143.24s)
    --- PASS: TestGrafana/GrafanaInstallVersionManifest (15.06s)
PASS
==================================================
--- PASS:
    GrafanaInstall [XXX XXX]
    GrafanaInstallVersionManifest [XXX XXX]
--- SKIP:
--- FAIL:
==================================================

DONE 3 tests in 158.637s
Tests passed
```

nikolaypopov commented on pkg/util/test/util.go:
```diff
@@ -2951,6 +2956,172 @@ func ValidatePrometheus(pxImageList map[string]string, cluster *corev1.StorageCl
+	shouldBeInstalled := cluster.Spec.Monitoring != nil && cluster.Spec.Monitoring.Grafana != nil &&
+		cluster.Spec.Monitoring.Grafana.Enabled && cluster.Spec.Monitoring.Prometheus != nil &&
+		cluster.Spec.Monitoring.Prometheus.Enabled
+	image := cluster.Status.DesiredImages.Grafana
```
I would recommend to follow how we validate all expected images, from `pxImageList`, because that is what we want to be installed, because `DesiredImages` is what PX Operator decided to install, but it doesn't necessary match which what it should be installing. The correct images should come versions manifest, wether its `px-versions` configmap or URL endpoint

nikolaypopov replied:
Just pass the `pxImageList` to the function, find your Grafana image there, which should be `pxImageList["grafana"]`, this should give you the value of it like so `grafana: docker.io/grafana/grafana:7.5.17`, then use the `validateContainerImageInsidePods()` function, because it will correctly trim the image to the correct URN, incase of custom registry or anything like that

nikolaypopov commented on pkg/util/test/util.go:
```diff
@@ -2951,6 +2955,177 @@ func ValidatePrometheus(pxImageList map[string]string, cluster *corev1.StorageCl
+	return nil
+}
+
+func ValidateGrafanaDeploymentImage(image string) error {
+	check := func() (interface{}, bool, error) {
+		pods, err := coreops.Instance().GetPods("kube-system", map[string]string{"app": "grafana"})
+		if err != nil {
+			return "", true, err
+		}
+
+		podCount := len(pods.Items)
+		if podCount > 0 && len(pods.Items[0].Spec.Containers) > 0 {
+			actualImage := pods.Items[0].Spec.Containers[0].Image
+			if actualImage == image {
+				logrus.Infof("Grafana installed successfully with image %s", image)
+				return "", false, nil
+			} else {
+				logrus.Errorf("Grafana not yet ready with image %s", image)
+				return "", true, fmt.Errorf("grafana is not installed with expected image %s. actual: %s", image, actualImage)
+			}
+		} else {
+			return "", true, fmt.Errorf("grafana is not installed with image %s. number of pods: %v", image, podCount)
+		}
+	}
+
+	if _, err := task.DoRetryWithTimeout(check, 1*time.Minute, 10*time.Second); err != nil {
+		return err
+	}
+
+	return nil
+}
```
Remove this since you using `validateContainerImageInsidePods()` and we don't need this any more

nikolaypopov commented on pkg/util/test/util.go:
```diff
@@ -2951,6 +2955,177 @@ func ValidatePrometheus(pxImageList map[string]string, cluster *corev1.StorageCl
+
+func ValidateGrafanaDeployment(cluster *corev1.StorageCluster, shouldBeInstalled bool, pxImageList map[string]string) error {
+	check := func() (interface{}, bool, error) {
+		pods, err := coreops.Instance().GetPods("kube-system", map[string]string{"app": "grafana"})
```
Don't hardcode `kube-system`, this can be installed into any namespace, let's get namespace from `cluster.Namespace` and use that instead

ggriffiths replied:
done

nikolaypopov commented on pkg/util/test/util.go:
```diff
@@ -2951,6 +2955,177 @@ func ValidatePrometheus(pxImageList map[string]string, cluster *corev1.StorageCl
+			return nil, true, err
+		}
+		if shouldBeInstalled {
+			if len(pods.Items) == 0 || pods.Items[0].Status.Phase != v1.PodRunning {
+				logrus.Errorf("Grafana not yet ready")
+				return "", true, fmt.Errorf("grafana is not installed when it should be")
+			}
```
Is grafana a deployment? Maybe just using `sched-ops` function `ValidateDeployment` should be enough here as it would do all the pod validation, if its not and its just a pod, this should be ok

nikolaypopov commented on pkg/util/test/util.go:
```diff
@@ -2951,6 +2955,177 @@ func ValidatePrometheus(pxImageList map[string]string, cluster *corev1.StorageCl
+
+			err := validateContainerImageInsidePods(cluster, pxImageList["grafana"], "grafana", pods)
+			if err != nil {
+				logrus.Errorf("Grafana image is invalid yet ready")
+				return "", true, fmt.Errorf("grafana is not installed when it should be")
```
Let's not log error and then return it, let's just return verbose error instead

nikolaypopov commented on pkg/util/test/util.go:
```diff
@@ -2951,6 +2955,177 @@ func ValidatePrometheus(pxImageList map[string]string, cluster *corev1.StorageCl
+				logrus.Infof("Grafana uninstalled successfully")
+				return "", false, nil
+			} else {
+				logrus.Errorf("Grafana not yet uninstalled")
+				return "", true, fmt.Errorf("grafana is installed when it is expected to be uninstalled")
```
Same here and all other place you have this

nikolaypopov reviewed (commented):
Overall looks much better, please address comments and build and image so I can run it in one of the basic jobs

nikolaypopov reviewed (approved):
LGTM, let's run `op-basic` job once with this private image before merging. Let me know when you have an image ready

ggriffiths commented:
Thanks, sent the test image.

ggriffiths commented:
Running here:
https://jenkins.pwx.dev.purestorage.com/job/Operator/view/Operator%20Release%20Dashboard/job/op-basic-ocp/209
